This repository comprises code to support results derived using UK Biobank approved project 15147 that have been published in:

Dand _et al_, GWAS meta-analysis of psoriasis identifies new susceptibility alleles impacting disease mechanisms and therapeutic targets, _MedRxiv_ (2023); [doi: 10.1101/2023.10.04.23296543](https://doi.org/10.1101/2023.10.04.23296543)

### Study lay summary

Psoriasis is a chronic inflammatory skin disease. We know that genetics plays an important role in who will get the disease, but many DNA sequence variants that influence risk remain to be found. We conducted a meta-analysis that combines data from 18 association studies, each of which looks at differences in DNA sequence between people with and without psoriasis, across the whole genome. Three of these studies were derived from large studies such as the UK Biobank, with the remaining 15 comprising research studies in Europe and North America into which psoriasis patients were recruited by specialist clinicians. In total, we studied around 36,000 people with psoriasis and 458,000 without.

We found 109 different regions of the genome (loci) that affect whether someone might develop psoriasis. Of these, 45 had not been previously identified. Genes at two of the new loci encode the proteins targeted by existing psoriasis drugs, giving potential insight into why these treatments are effective. To better understand the immune mechanisms that cause psoriasis and highlight potential avenues for new treatment development, we focused on DNA sequence variants in protein-coding genes that are predicted to alter protein structure, and others that may influence how much of each gene is expressed in a person's skin or blood. We also cross-referenced our findings against data from single-cell expression experiments, which highlighted specific immune cell types in the skin within which our DNA sequence variants are likely to be exerting their effect on psoriasis risk.

### Data

UK Biobank data cannot be openly shared, but researchers can apply for access.

The presence of psoriasis was established from participant self-report (field 20002), self-reported medication (field 20003), hospital episode statistic summary data (fields 41202 and 41204) and record-level primary care clinical event (diagnosis) data (field 42040). Definitions of cases, controls and excluded participants based on these fields are described in the table below. Psoriasis phenotype values have been returned to UK Biobank.

GWAS were performed using genome-wide imputed genotype data (field 22828). Participant QC exclusion lists (which include non-"white British" participants and related individuals) and covariate values (principal components and array type) were derived from the centrally generated QC summary variables (see [resource 531](https://biobank.ndph.ox.ac.uk/showcase/refer.cgi?id=531)) as detailed in the scripts linked below. Additional participant exclusions were identified using minimal lists of markers in common with other studies contributing to the meta-analysis, and their associated genotype data. These data can not be made publicly available.

### Code

Extension **.R** for R scripts, **.sh** for bash scripts.
R scripts generally include a dependency on [this custom package](https://github.com/njd34/R_helpers) of utility functions.

| Script                                                                                                                                           | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
| ------------------------------------------------------------------------------------------------------------------------------------------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| _Generate psoriasis phenotype_                                                                                                                   |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| [r01_preprocess_data_step1only.R](r01_preprocess_data_step1only.R)                                                                               | Convert UKB data download from raw tabular data into an R data frame                                                                                                                                                                                                                                                                                                                                                                                                                   |
| [r06full_phenotyping_psoriasis.R](r06full_phenotyping_psoriasis.R)                                                                               | Derive a series of different psoriasis case lists based on four lines of evidence: self-reported disease, self-reported medication (list below), HES diagnosis, GP diagnosis (code list and bash script below).                                                                                                                                                                                                                                                                        |
| [final_meds_list_pso.txt](final_meds_list_pso.txt)                                                                                               | List of self-reported medications (field 20003) confirmed to be relevant to psoriasis by manual review by 2Ã— dermatologists. Based on medications reported by participants self-reporting psoriasis (field 20002). Manually generated from Excel                                                                                                                                                                                                                                       |
| [psoriasis_automapped_L40_v2.list](psoriasis_automapped_L40_v2.list)<br>[psoriasis_automapped_L40_CTV3.list](psoriasis_automapped_L40_CTV3.list) | List of primary care diagnosis codes (read_v2 and read_ctv3) corresponding to psoriasis, derived from the ICD-to-read code mappings in the UKB resource *all_lkps_maps.xlsx* ([resource 592](https://biobank.ndph.ox.ac.uk/ukb/refer.cgi?id=592), original version)                                                                                                                                                                                                                    |
| [extract_ids_by_pheno.sh](extract_ids_by_pheno.sh)                                                                                               | Generate a list of participant eids that match the provided v2 and ctv3 read code lists in the primary care table *gp_clinical.txt*                                                                                                                                                                                                                                                                                                                                                    |
| [r11_get_simple_case_ctrl_pheno.R](r11_get_simple_case_ctrl_pheno.R)                                                                             | Produce a psoriasis phenotype (.pheno) file for GWAS analysis, based on the case lists from script s06full (above). The chosen definitions were:<br><br>**Cases**: participants with two or more of self-reported psoriasis, self-reported relevant treatment, HES psoriasis diagnosis, GP psoriasis diagnosis<br><br>**Excluded**: participants with only one of self-reported psoriasis, HES psoriasis diagnosis, GP psoriasis diagnosis<br><br>**Controls**: all other participants |
| _QC genetic data and generate covariates_                                                                                                        |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| Available in github repo [ukb15147/central-scripts](https://github.com/ukb15147/central-scripts)                                                 | -                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
| _Identify inter-study related participants_                                                                                                      |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| [compare_genotypes.sh](compare_genotypes.sh)                                                                                                     | Outline steps for the identification of related participants between participating GWAS studies based on minimal common marker sets                                                                                                                                                                                                                                                                                                                                                    |
| [s05_compare_9new.sh](s05_compare_9new.sh)                                                                                                       | Example script for the dataset merge and relationship identification step in the above                                                                                                                                                                                                                                                                                                                                                                                                 |
| [s08_extract_exclusions_final.R](s08_extract_exclusions_final.R)                                                                                 | Identify participants to be excluded from each individual GWAS based on the relationships estimated above.                                                                                                                                                                                                                                                                                                                                                                             |
| _Run GWAS analysis_                                                                                                                              |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| [plink_gwas.sh](plink_gwas.sh)                                                                                                                   | Example PLINK2 code to run logistic regression GWAS                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| _Harmonise results to other studies_                                                                                                             |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| [harmonise.sh](harmonise.sh)                                                                                                                     | Format UKB GWAS summary statistics and ensure alleles and variant IDs are consistent with other meta-analysis input studies                                                                                                                                                                                                                                                                                                                                                            |
| _Run meta-analysis_                                                                                                                              |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| [s27_assemble_input_from_list_final.sh](s27_assemble_input_from_list_final.sh)                                                                   | Bash script to create final UK Biobank results file to be input into meta-analysis (and input files for all other studies). Based on GWAS sumstats output from **harmonise.sh** (above), plus meta-data (see below)                                                                                                                                                                                                                                                                    |
| [cohorts_with_sample_sizes_vExample.txt](cohorts_with_sample_sizes_vExample.txt)                                                                 | Example meta-data file for use with script s27                                                                                                                                                                                                                                                                                                                                                                                                                                         |
| [METAL_final_v4.config](METAL_final_v4.config)                                                                                                   | Config file used to run meta-analysis using METAL v2020-05-05 with command<br>`./metal METAL_final_v4.config`                                                                                                                                                                                                                                                                                                                                                                          |
 